### PR TITLE
feat(issues): add support for exporting issues to JSON

### PR DESCRIPTION
**Pull request by burntcarrot**
* add `--json` flag
* add `--output` flag
* add custom types for marshaling to JSON
* add helper utilities for JSON conversion

Signed-off-by: burntcarrot <aadhav.n1@gmail.com>

### CODE REVIEW

File: command/issues/list/list.go [hunk 1]
```diff
@@ -48,6 +62,13 @@ func NewCmdIssuesList() *cobra.Command {
 
 	// --limit, -l flag
 	cmd.Flags().IntVarP(&opts.LimitArg, "limit", "l", 30, "Fetch the issues upto the specified limit")
+
+	// --output, -o flag
+	cmd.Flags().StringVarP(&opts.OutputArg, "output", "o", "", "Filename for exporting issues")
```
Comment (burntcarrot) [inline command/issues/list/list.go:67]: Will convert `--output` to `--output-file`.

File: command/issues/list/list.go [hunk 2]
```diff
@@ -48,6 +62,13 @@ func NewCmdIssuesList() *cobra.Command {
 
 	// --limit, -l flag
 	cmd.Flags().IntVarP(&opts.LimitArg, "limit", "l", 30, "Fetch the issues upto the specified limit")
+
+	// --output, -o flag
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go]:
```suggestion
	// --output-file, -o flag
```

File: command/issues/list/list.go [hunk 3]
```diff
@@ -48,6 +62,13 @@ func NewCmdIssuesList() *cobra.Command {
...
+	cmd.Flags().StringVarP(&opts.OutputArg, "output-file", "o", "", "Filename for exporting issues")
+
+	// --json flag
+	cmd.Flags().BoolVar(&opts.JSONArg, "json", false, "Export issues to JSON")
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:70]:
```suggestion
	cmd.Flags().BoolVar(&opts.JSONArg, "json", false, "Output reported issues in JSON format")
```

File: command/issues/list/list.go [hunk 4]
```diff
@@ -48,6 +62,13 @@ func NewCmdIssuesList() *cobra.Command {
 
 	// --limit, -l flag
 	cmd.Flags().IntVarP(&opts.LimitArg, "limit", "l", 30, "Fetch the issues upto the specified limit")
+
+	// --output, -o flag
+	cmd.Flags().StringVarP(&opts.OutputArg, "output-file", "o", "", "Filename for exporting issues")
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:67]:
```suggestion
	cmd.Flags().StringVarP(&opts.OutputArg, "output-file", "o", "", "Output file to write the reported issues to")
```

File: command/issues/list/list.go [hunk 5]
```diff
@@ -80,7 +101,19 @@ func (opts *IssuesListOptions) Run() (err error) {
...
 	// Fetch the list of issues using SDK based on user input
 	ctx := context.Background()
-	return opts.getIssuesData(ctx)
+	err = opts.getIssuesData(ctx)
+	if err != nil {
+		return err
+	}
+
+	// if --json is passed, export issue data to JSON
+	if opts.JSONArg {
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:110]: Try using switch case here? Gives more clarity.

File: command/issues/list/list.go [hunk 6]
```diff
@@ -135,3 +166,73 @@ func (opts *IssuesListOptions) showIssues() {
 	// Using pterm to render the list of list
 	pterm.DefaultTable.WithSeparator("\t").WithData(opts.ptermTable).Render()
 }
+
+// exportJSON exports issues as JSON
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:170]:
```suggestion
// Handles exporting issues as JSON
```

File: command/issues/list/list.go [hunk 7]
```diff
@@ -135,3 +166,73 @@ func (opts *IssuesListOptions) showIssues() {
...
+			return err
+		}
+
+		pterm.Info.Printf("Saved issues to %s!\n", filename)
+	}
+
+	return nil
+}
+
+// convertJSON converts issueData to a JSON-compatible struct
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:197]:
```suggestion
// Converts issueData to a JSON-compatible struct
```

File: command/issues/list/list.go [hunk 8]
```diff
@@ -18,11 +20,23 @@ type IssuesListOptions struct {
 	FileArg        string
 	RepoArg        string
 	LimitArg       int
+	OutputArg      string
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:23]:
```suggestion
	OutputFilenameArg      string
```

File: command/issues/list/list.go [hunk 9]
```diff
@@ -80,7 +101,19 @@ func (opts *IssuesListOptions) Run() (err error) {
...
 	ctx := context.Background()
-	return opts.getIssuesData(ctx)
+	err = opts.getIssuesData(ctx)
+	if err != nil {
+		return err
+	}
+
+	// if --json is passed, export issue data to JSON
+	if opts.JSONArg {
+		opts.exportJSON(opts.OutputArg)
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:111]:
```suggestion
		opts.exportJSON(opts.OutputFilenameArg)
```

File: command/issues/list/list.go [hunk 10]
```diff
@@ -135,3 +166,73 @@ func (opts *IssuesListOptions) showIssues() {
...
+	if filename == "" {
+		pterm.Println(string(data))
+	} else {
+		err = ioutil.WriteFile(filename, data, 0644)
+		if err != nil {
+			return err
+		}
+
+		pterm.Info.Printf("Saved issues to %s!\n", filename)
+	}
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:192]: We can write this in a cleaner way using guard clause:
```go
func exportJSON(...) err error {
  ...
  if filename == "" {
		pterm.Println(string(data))
		return nil
  }

  if err = ioutil.WriteFile(filename, data, 0644); err != nil {
			return err
  }
  pterm.Info.Printf("Saved issues to %s!\n", filename)
  return nil
}
```

Also, are we handling the case in which the user doesn't explicitly specify a filename for the JSON file? In that case we should export the issues to `deepsource_issues.json` file.

File: command/issues/list/list.go [hunk 11]
```diff
@@ -135,3 +166,73 @@ func (opts *IssuesListOptions) showIssues() {
...
+
+	return nil
+}
+
+// convertJSON converts issueData to a JSON-compatible struct
+func convertJSON(issueData []issues.Issue) (ExportData, error) {
```
Comment (siddhant-deepsource) [inline command/issues/list/list.go:198]: Are we returning `error` anywhere here?

**Review by siddhant-deepsource — CHANGES_REQUESTED** [review]
This looks good and almost ready. Have made some comments though and requested some changes. Please take a look and address them. Also, let me know in case you have any doubts/queries regarding them.

File: command/issues/list/list.go (hunk 10 above)
Comment (burntcarrot) [reply command/issues/list/list.go:192]: Cobra doesn't allow blank filenames when `--output-file` is passed. So, the user needs to pass a value while specifying the `--output-file` flag.

File: command/issues/list/list.go (hunk 11 above)
Comment (burntcarrot) [reply command/issues/list/list.go:198]: Nice catch! I forgot to refactor it :P

File: command/issues/list/list.go (hunk 10 above)
Comment (siddhant-deepsource) [reply command/issues/list/list.go:192]: Cool, what error does it display when the filename is not passed?

Comment (burntcarrot) [reply command/issues/list/list.go:192]: `ERROR` flag needs an argument: --output-file

File: command/issues/list/list.go (hunk 5 above)
Comment (siddhant-deepsource) [reply command/issues/list/list.go:110]: Ignore.

**Pull request merged**